### PR TITLE
Improve Blazor project template

### LINKED ISSUE #52079: Enabled option "Do not use top-level statements" ignored in "Blazor Web App" template for WebAssembly client
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Checked option "Do not use top-level statements" ignored in "Blazor Web App" template for WebAssembly client. The server side will be created without top-level statements but the WeAssembly client is created **with** top-level statements regardless of option selected.

### Expected Behavior

I would expect both the client and server projects to not use top-level statements. When that option is checked.

### Steps To Reproduce

Create a new "Blazor Web App" project and select "Do not use top-level statements" then click create. The server project will not use top-level statements but the client project will unexpectedly use top-level statements.

### Exceptions (if any)

_No response_

### .NET Version

8.0.100

### Anything else?

I am not positive that this is the repo to report this particular issue. I could not find this issue reported anywhere, but I may be missing it.

### LINKED ISSUE #52084: BWA project template: "@" on @render-mode values and `.Client` project assembly discovery
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Creating an app from the BWA project template with global WASM interactivity produces ...

```razor
<HeadOutlet @rendermode="@InteractiveWebAssembly" />
...
<Routes @rendermode="@InteractiveWebAssembly" />
```

Although the at-symbols (`@`) on the values are fine, it departs from our guidance to not place them on values of attribute directives outside of explicit Razor expressions [`@(...)`].

I recommend 🔪 them off; but if you plan to keep these, it creates a delta between the framework and docs, so how would you like to proceed? Let the delta stand, or change all of the docs to this format for directive attributes? If the latter, it will need to be pushed back quite a bit (24Q1/Q2) because it's a 🐘 job to change them everywhere and I'm 🏃‍♂️ on many high priority issues for quite a while.

I opened a docs issue just to call out that this is a supported approach if the dev wants to write their markup this way. https://github.com/dotnet/AspNetCore.Docs/issues/31042

Second ❓: WRT the logic for finding the `.Client` project assembly ...

```
#if (UseWebAssembly && SampleContent)
    .AddAdditionalAssemblies(typeof(Counter).Assembly);
#elif (UseWebAssembly)
    .AddAdditionalAssemblies(typeof(BlazorWeb_CSharp.Client._Imports).Assembly);
#endif
```

I don't see why the latter `_Imports` file-based approach isn't adopted even if sample content is present. After all, even the empty solution (no sample pages) has an `_Imports` file. As soon as the dev 🔪 out the `Counter` component because any production app isn't going to keep it, it will :boom:. Many new devs and new-to-Blazor devs won't know the second approach for the `_Imports` file off the top of their head.

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Exceptions (if any)

_No response_

### .NET Version

_No response_

### Anything else?

_No response_

### LINKED ISSUE #52167: Add Account/AccessDenied endpoint to blazor web template
This was reported by a customer at https://devblogs.microsoft.com/dotnet/announcing-asp-net-core-in-dotnet-8/comment-page-2/#comment-20005

> Thank you for your amazing work with the new framework, and especially Blazor!
> I’m setting up authorization in my BlazorApp. When I required a specific policy for a page, and the logged in user isn’t in a role in that policy, I’m redirected to Account/AccessDenied, as expected. But this page is missing in the template. I created it manually, but I guess you would want to add it to the template in the future.

This is an oversight. "/Account/AccessDenied" is the default `CookieAuthenticationOptions.AccessDeniedPath` which is is used by our template, and is [part of the Identity UI razor pages.](https://github.com/dotnet/aspnetcore/blob/76b5785b4ca4d6107baad792e2fe4d7b6f938b9b/src/Identity/UI/src/Areas/Identity/Pages/V5/Account/AccessDenied.cshtml)

### PR DESCRIPTION
**Pull request by halter73**
# Use consistent code style in Blazor project templates

- Removed usage of top-level statements in client project if "Do not use top-level statements" is selected
- Removed extra "@" in `@render-mode` values
- Always use `typeof(Namespace._Imports).Assembly` instead of `typeof(Counter).Assembly` so the compilation does not break when the sample Counter component is removed, and so the code is more consistent with how it is when no sample content is generated
- Added Account/AccessDenied endpoint to individual auth option to match Identity UI razor pages. This is shown when the user is authenticated but unauthorized by default.

Fixes #52079
Fixes #52084
Fixes #52167

## Customer Impact

In addition to not using top-level statements when the customer requests that we don't, this improves code style consistency within the Blazor project template and with the Blazor docs.

## Regression?

- [ ] Yes
- [x] No

## Risk

- [ ] High
- [ ] Medium
- [x] Low

These are small stylistic changes to the Blazor project templates.

## Verification

- [x] Manual (required)
- [ ] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

### CODE REVIEW

**Comment by halter73** [comment]
@guardrex Regarding the following from #51084

> Second ❓: WRT the logic for finding the `.Client` project assembly ...
>
> ```
> #if (UseWebAssembly && SampleContent)
>     .AddAdditionalAssemblies(typeof(Counter).Assembly);
> #elif (UseWebAssembly)
>    .AddAdditionalAssemblies(typeof(BlazorWeb_CSharp.Client._Imports).Assembly);
> #endif
> ```
>
> I don't see why the latter `_Imports` file-based approach isn't adopted even if sample content is present. After all, even the empty solution (no sample pages) has an `_Imports` file. As soon as the dev 🔪 out the `Counter` component because any production app isn't going to keep it, it will :boom:. Many new devs and new-to-Blazor devs won't know the second approach for the `_Imports` file off the top of their head.

I don't have a huge preference one way or the other. I think @SteveSandersonMS chose `Counter` initially, and then I later added `_Imports` for the non-sample case. I left it as `Counter` for the default case because it does look better, but if everyone is okay with changing it to always use `_Imports`, I can do that as part of this PR.

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp.Client/Program.Main.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#if (IndividualLocalAuth)
+using BlazorWeb_CSharp.Client;
+using Microsoft.AspNetCore.Components.Authorization;
+#endif
+using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
+
+namespace BlazorWeb_CSharp.Client;
+
+class Program
```
Comment (halter73) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp.Client/Program.Main.cs:9]: Unlike the server template, this does not have an accessibility modifier. Neither does the `Main` method. I would have made both the class and `public` like we do for the server project, but this made the `@typeof(Program).Assembly` reference in Routes.razor ambiguous.

https://github.com/dotnet/aspnetcore/blob/d8acc61aa53bbc5ed18749677d2afa6bd38fcc31/src/ProjectTemplates/Web.ProjectTemplates/content/BlazorWeb-CSharp/BlazorWeb-CSharp/Components/Routes.razor#L5

I could change that to be `@typeof(BlazorWeb_CSharp.Program).Assembly` instead and make this public. And then only qualify it when you've disabled top level statements.

Does anyone prefer that, or is everyone fine the way it is?

**Comment by danroth27** [comment]
@halter73 Can you update Routes.razor as well to remove the unnecessary leading `@` chars?:

```razor
<Router AppAssembly="typeof(Program).Assembly">
    <Found Context="routeData">
        <RouteView RouteData="routeData" DefaultLayout="typeof(Layout.MainLayout)" />
        <FocusOnNavigate RouteData="routeData" Selector="h1" />
    </Found>
</Router>
```

**Comment by guardrex** [comment]
@danroth27, we discussed the delta that removing those `@` will create with the docs. If you go without them in the project template, shouldn't I make the doc set match? I'd kind'a prefer to do it because readers will spot it and open issues (and even patch PRs) for what they think are doc bugs.

**Comment by guardrex** [comment]
@halter73 ... I think it will break new devs/new Blazor devs fairly fast, but I'm opening a doc issue for our `AdditionalAssemblies` section at ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#route-to-components-from-multiple-assemblies

... to cover the situation. I'll specifically call out the situation for BWAs in versioned content.

**Comment by danroth27** [comment]
> @danroth27, we discussed the delta that removing those `@` will create with the docs. If you go without them in the project template, shouldn't I make the doc set match? I'd kind'a prefer to do it because readers will spot it and open issues (and even patch PRs) for what they think are doc bugs.

@guardrex Yeah, we're trying to standardize on a consistent style here. The style we're going with is to only include the `@` if it's needed.

**Review by javiercn — APPROVED** [review]

**Comment by guardrex** [comment]
@halter73 @javiercn ... Passing note and ❓ before we go here.

These definitions are similar WRT locating components ...

`AdditionalAssemblies` param on the Blazor router instance ...

> Gets or sets a collection of additional assemblies that should be searched for components that can match URIs.

`AddAdditionalAssemblies` method for the endpoint convention builder ...

Summary ...

> Adds the given additional assemblies to the component application.

Remark ...

> The provided assemblies will be scanned for pages that will be mapped as endpoints.

Two concerns ...

1. I think it's a good idea if I write that the `AddAdditionalAssemblies` method is required for client-side components in Blazor Web Apps without regard to whether or not components in the `.Client` project are routable or not. It's required for general component discovery by the framework (i.e., outside of routing). If it isn't there, routable client-side components throw a 404 💥.

2. That remark for the `AddAdditionalAssemblies` method kind'a makes it sound like one wouldn't need to specify the `AdditionalAssemblies` param on the router in order to discover routable client-side components. On a lark, I tried removing the param from the router, leaving just the `AddAdditionalAssemblies` method on `app.MapRazorComponents<App>()` and ***it worked*** and found the `Counter` component in the `.Client` project. What should we say about this? Should we just say in a BWA that ***both*** should be present anyway? ... which is what the project template sets up by default.

**Comment by danroth27** [comment]
@guardrex I believe you need to specify `AddAdditionalAssemblies` on the `Router` to discover components in other assemblies when the `Router` is set up to be interactive, either from the server or the client. If the `Router` doesn't have an interactive render mode, then `AddAdditionalAssemblies` on `MapRazorComponents` should be sufficient.

**Comment by guardrex** [comment]
@danroth27 ... Thanks. I understand. I'll make further updates shortly and ping for review on the PR.

**Comment by guardrex** [comment]
Ok ... I think I get it now. I'll make the updates the PR and ping for review.

**Pull request merged**